File: c++/nitf/tests/test_image_loading.cpp
/* =========================================================================
 * This file is part of NITRO
 * =========================================================================
 *
 * (C) Copyright 2004 - 2010, General Dynamics - Advanced Information Systems
 *
 * NITRO is free software; you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this program; if not, If not,
 * see <http://www.gnu.org/licenses/>.
 *
 */

#include <import/nitf.hpp>
#include <import/except.h>

void writeImage(nitf::ImageSegment &segment,
                nitf::Reader &reader,
                const int imageNumber,
                const char *imageName,
                nitf_Uint32 rowSkipFactor,
                nitf_Uint32 columnSkipFactor, bool optz)
{
    size_t subWindowSize;
    nitf::SubWindow subWindow;
    unsigned int i;
    int padded;
    nitf::Uint8** buffer = NULL;
    nitf::Uint32 band;
    nitf::Uint32 * bandList;

    nitf::ImageReader deserializer = reader.newImageReader(imageNumber);

    // missing skip factor
    nitf::ImageSubheader subheader = segment.getSubheader();

    nitf::Uint32 nBits   = subheader.getNumBitsPerPixel();

    nitf::Uint32 nBands  = subheader.getNumImageBands();
    nitf::Uint32 xBands  = subheader.getNumMultispectralImageBands();
    nBands += xBands;


    nitf::Uint32 nRows   = subheader.getNumRows();
    nitf::Uint32 nCols   = subheader.getNumCols();
    subWindowSize = (size_t)(nRows / rowSkipFactor) *
        (size_t)(nCols / columnSkipFactor) *
        (size_t)NITF_NBPP_TO_BYTES(nBits);

    if (optz)
    {
        std::string irep =
                segment.getSubheader().getImageRepresentation().toString();
        std::string ic =
                segment.getSubheader().getImageCompression().toString();

        if (nBands == 3 && segment.getSubheader().getImageMode().toString()
                == "P" && str::startsWith(irep, "RGB")
                && NITF_NBPP_TO_BYTES(nBits) == 1 && str::startsWith(ic, "N"))
        {
            subWindowSize *= nBands;
            nBands = 1;
            std::cout << "Using accelerated 3-band RGB mode pix-interleaved image" << std::endl;
        }
        if (nBands == 2 && segment.getSubheader().getImageMode().toString()
                == "P" && str::startsWith(ic, "N"))
        {

            subWindowSize *= nBands;
            nBands = 1;
            std::cout << "Using accelerated 2-band IQ mode pix-interleaved image" << std::endl;
        }
    }


    std::cout << "NBANDS -> " << nBands << std::endl;
    std::cout << "XBANDS -> " << xBands << std::endl;
    std::cout << "NROWS -> "  << nRows  << std::endl;
    std::cout << "NCOLS -> "  << nCols  << std::endl;
    std::cout << "PVTYPE -> " << subheader.getPixelValueType().toString() << std::endl;
    std::cout << "NBPP -> " << subheader.getNumBitsPerPixel() .toString() << std::endl;
    std::cout << "ABPP -> " << subheader.getActualBitsPerPixel().toString() << std::endl;
    std::cout << "PJUST -> " << subheader.getPixelJustification().toString() << std::endl;
    std::cout << "IMODE -> " << subheader.getImageMode().toString() << std::endl;
    std::cout << "NBPR -> " << subheader.getNumBlocksPerRow().toString() << std::endl;
    std::cout << "NBPC -> " << subheader.getNumBlocksPerCol().toString() << std::endl;
    std::cout << "NPPBH -> " << (int)subheader.getNumPixelsPerHorizBlock() << std::endl;
    std::cout << "NPPBV -> " << (int)subheader.getNumPixelsPerVertBlock() << std::endl;
    std::cout << "IC -> " << subheader.getImageCompression().toString() << std::endl;
    std::cout << "COMRAT -> " << subheader.getCompressionRate().toString() << std::endl;

    std::cout << "Allocating work buffer..." << std::endl;

    buffer = new nitf::Uint8*[nBands];
    assert(buffer);

    band = 0;
    bandList = new nitf::Uint32[nBands];

    subWindow.setStartCol(0);
    subWindow.setStartRow(0);

    subWindow.setNumRows(nRows / rowSkipFactor);
    subWindow.setNumCols(nCols / columnSkipFactor);

    nitf::PixelSkip pixelSkip(rowSkipFactor, columnSkipFactor);

    subWindow.setDownSampler(&pixelSkip);

    for (band = 0; band < nBands; band++)
    {
        bandList[band] = band;
        buffer[band] = new nitf::Uint8[subWindowSize];
        assert(buffer[band]);
    }
    subWindow.setBandList(bandList);
    subWindow.setNumBands(nBands);

    std::cout << "Reading image..." << std::endl;
    deserializer.read(subWindow, buffer, &padded);

    std::cout << "Call completed!" << std::endl;

    std::cout << "Writing bands..." << std::endl;
    for (i = 0; i < nBands; i++)
    {
        std::cout << "Writing band # " << i << std::endl;
        std::string base = sys::Path::basename(imageName);

        size_t where = 0;
        while ((where = base.find(".")) != (size_t)std::string::npos)
            base.replace(where, 1, "_");

        std::ostringstream file;
        file << base << "__" << imageNumber << "__"
             << nRows / rowSkipFactor << '_'
             << nCols / columnSkipFactor << '_'
             << nBits << "_band_" << i << ".out";

        nitf::IOHandle toFile(file.str(), NITF_ACCESS_WRITEONLY, NITF_CREATE);
        toFile.write((const char *) buffer[i], subWindowSize);
        toFile.close();
        std::cout << "Finished # " << i << std::endl;
    }

    for (band = 0; band < nBands; band++)
        delete [] buffer[band];
    delete [] buffer;
    delete [] bandList;
}


int main(int argc, char **argv)
{
    /* Skip factors */
    nitf_Uint32 rowSkipFactor = 1;
    nitf_Uint32 columnSkipFactor = 1;

    try
    {
        /*  If you didnt give us a nitf file, we're croaking  */
        if (argc < 2)
        {
            std::cout << "Usage: " << argv[0] << " <nitf-file> [-o]"
                      << std::endl;
            exit(EXIT_FAILURE);
        }

        bool optz = (argc > 2 && std::string(argv[2]) == "-o");

        /*  This is the reader  */
        nitf::Reader reader;

        /*  This is the io handle we will give the reader to parse  */
        nitf::IOHandle io(argv[1]);

        /*  Read the file (first pass) */
        nitf::Record record = reader.read(io);
        /*  These iterators are for going through the image segments  */
        nitf::ListIterator iter;
        nitf::ListIterator end;

        /*  And set this one to the end, so we'll know when we're done!  */
        iter = record.getImages().begin();
        end = record.getImages().end();
        std::cout << "Iterating list of images..." << std::endl;
        /*  While we are not done...  */
        for (int count = 0; iter != end; ++iter, ++count)
        {
            std::cout << "Getting image segment..." << std::endl;
            nitf::ImageSegment imageSegment((nitf_ImageSegment *) iter.get());

            std::cout << "Retrieved." << std::endl;
            std::cout << "Writing image... " << std::endl;

            /*  Write the thing out  */
            writeImage(imageSegment, reader, count, argv[1],
                       rowSkipFactor, columnSkipFactor, optz);
            std::cout << "done." << std::endl;
        }
    }
    catch (except::Exception &thr)
    {
        std::cout << "Caught module exception: " << thr.toString() << std::endl;
    }
    catch (std::exception & ex)
    {
        std::cout << "Caught C++ exception: " << ex.what() << std::endl;
    }
    catch (...)
    {
        std::cout << "Caught unknown exception" << std::endl;
    }

    return 0;
}

File: externals/nitro/modules/java/nitf/src/samples/TestImageLoading.java
/* =========================================================================
 * This file is part of NITRO
 * =========================================================================
 * 
 * (C) Copyright 2004 - 2010, MDA Information Systems LLC
 *
 * NITRO is free software; you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public 
 * License along with this program; if not, If not, 
 * see <http://www.gnu.org/licenses/>.
 *
 */

import java.io.File;
import java.util.Arrays;
import java.util.Iterator;
import java.util.List;
import java.util.Vector;

import nitf.DownSampler;
import nitf.IOHandle;
import nitf.ImageReader;
import nitf.ImageSegment;
import nitf.NITFException;
import nitf.PixelSkipDownSampler;
import nitf.Reader;
import nitf.Record;
import nitf.SubWindow;

/**
 * This test reads an image and writes out the raw image data You must have
 * write privileges in the directory this is run in, or you can pass in as an
 * optional parameter an output directory. Also, you can pass in a row and
 * column skip factor, which can be used to downsample the image (decimation).
 */
public class TestImageLoading
{

    /**
     * Performs the write call, writing one row at a time
     * 
     * @param segment
     *            the image sement to write
     * @param outputDir
     *            the output directory
     * @param baseName
     *            the base name of the input image
     * @param deserializer
     *            the imagereader object
     * @param imageNumber
     *            the image number in the file
     */
    private static void writeImage(ImageSegment segment, File outputDir,
            String baseName, ImageReader deserializer, int imageNumber,
            DownSampler downSampler)
    {

        int nBits = segment.getSubheader().getNumBitsPerPixel().getIntData();
        int nBands = segment.getSubheader().getNumImageBands().getIntData();
        int xBands = segment.getSubheader().getNumMultispectralImageBands()
                .getIntData();
        // get the number of bands
        nBands += xBands;

        int nRows = segment.getSubheader().getNumRows().getIntData();
        int nColumns = segment.getSubheader().getNumCols().getIntData();

        int rowSkip = 1;
        int colSkip = 1;

        if (downSampler != null)
        {
            rowSkip = downSampler.getRowSkip();
            colSkip = downSampler.getColSkip();
            nRows /= rowSkip;
            nColumns /= colSkip;
        }

        int subimageSize = nColumns * ((nBits - 1) / 8 + 1);

        byte[][] buffer = new byte[nBands][subimageSize];

        try
        {
            IOHandle[] handles = new IOHandle[nBands];

            // first, create an output handle for each band
            for (int i = 0; i < nBands; ++i)
            {
                File outFile = new File(outputDir, baseName + "_" + nRows + "x"
                        + nColumns + "_" + imageNumber + "_BAND" + i + ".out");

                System.out.println("Creating output file: "
                        + outFile.getAbsolutePath());

                // create an output write-only handle
                IOHandle handle = new IOHandle(outFile.getPath(),
                        IOHandle.NITF_ACCESS_WRITEONLY, IOHandle.NITF_CREATE);

                handles[i] = handle;
            }

            // set the band list
            int[] bands = new int[nBands];
            for (int i = 0; i < bands.length; i++)
                bands[i] = i;

            // set up the subwindow
            SubWindow subimage = new SubWindow();
            subimage.setStartCol(0);
            subimage.setNumCols(nColumns);
            subimage.setNumRows(1);
            subimage.setBandList(bands);
            subimage.setNumBands(nBands);

            // read one row at a time, and write it to the file
            // Note- this is not very efficient at all
            for (int i = 0; i < nRows; ++i)
            {
                subimage.setStartRow(i * rowSkip);

                if (downSampler != null)
                    subimage.setDownSampler(downSampler);

                // read the image data to the buffer
                boolean status = deserializer.read(subimage, buffer);

                // iterate over the bands, and write the data to its own file
                for (int j = 0; j < nBands; ++j)
                {
                    // write the data to the output handle
                    handles[j].write(buffer[j], subimageSize);
                }
            }

            // now close each handle
            for (int i = 0; i < nBands; ++i)
            {
                handles[i].close();
                System.out.println("Done writing to file: "
                        + handles[i].getFileName());
            }
        }
        catch (NITFException nitx)
        {
            nitx.printStackTrace();
            return;
        }
    }

    public static void main(String[] args) throws NITFException
    {
        // must supply a filename
        if (args.length < 1)
        {
            System.err
                    .println("usage: java TestImageLoading fileName [-sr skipRows] [-sc skipCols] [outputDir]");
            System.exit(0);
        }

        List argList = new Vector();
        argList.addAll(Arrays.asList(args));
        argList.remove(0);

        int skipRows = 1;
        int skipCols = 1;
        // by default, set the output to the current dir
        String outDirname = System.getProperty("user.dir");

        for (Iterator iter = argList.iterator(); iter.hasNext();)
        {
            String arg = (String) iter.next();
            if (arg.equals("-sr") && iter.hasNext())
            {
                skipRows = Integer.parseInt((String) iter.next());
            }
            else if (arg.equals("-sc") && iter.hasNext())
            {
                skipCols = Integer.parseInt((String) iter.next());
            }
            else
                outDirname = arg;
        }

        // normalize, if bad data
        skipRows = skipRows <= 0 ? 1 : skipRows;
        skipCols = skipCols <= 0 ? 1 : skipCols;

        // try to get the outputDir, if one is passed in
        File outputDir = new File(outDirname);
        outputDir.mkdirs();

        // must be able to write to it, or exit
        if (!outputDir.canWrite())
        {
            System.err.println("Unable to write to output directory ["
                    + outputDir + "]");
            System.exit(1);
        }

        // create a new nitf.Reader object
        Reader reader = new Reader();

        // create an IOHandle to open the file read-only
        IOHandle handle = new IOHandle(args[0], IOHandle.NITF_ACCESS_READONLY,
                IOHandle.NITF_OPEN_EXISTING);

        // read the file, and get back the nitf.Record object
        Record record = reader.read(handle);

        // make a downsampler
        DownSampler downSampler = new PixelSkipDownSampler(skipRows, skipCols);

        // get the image segments
        ImageSegment[] segments = record.getImages();
        for (int i = 0; i < segments.length; ++i)
        {
            ImageSegment segment = segments[i];

            // get a new ImageReader
            ImageReader deserializer = reader.getNewImageReader(i);

            // write the image
            writeImage(segment, outputDir, new File(args[0]).getName(),
                    deserializer, 0, downSampler);
        }

        // close the handle
        handle.close();
    }

}
